ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/sqlstd/SQLAuthorizationUtils.java

```diff
@@ -454,13 +456,24 @@ private static Set<SQLPrivTypeGrant> getPrivilegesFromFS(
     if (FileUtils.isOwnerOfFileHierarchy(fs, fileStatus, userName, recurse)) {
       privs.add(SQLPrivTypeGrant.OWNER_PRIV);
     }
-    if (FileUtils.isActionPermittedForFileHierarchy(fs, fileStatus, userName, FsAction.WRITE, recurse)) {
+    UserGroupInformation ugi = Utils.getUGI();
+    String currentUser = ugi.getShortUserName();
+    FileSystem fsAsUser = null;
+    UserGroupInformation proxyUser = null;
+    if (userName != null && !userName.equals(currentUser)) {
+      proxyUser = UserGroupInformation.createProxyUser(userName, UserGroupInformation.getLoginUser());
+      fsAsUser = FileUtils.getFsAsUser(fs, proxyUser);
+    }
+    if (FileUtils.isActionPermittedForFileHierarchy(fs, fileStatus, userName, FsAction.WRITE, recurse, fsAsUser)) {
       privs.add(SQLPrivTypeGrant.INSERT_NOGRANT);
       privs.add(SQLPrivTypeGrant.DELETE_NOGRANT);
     }
-    if (FileUtils.isActionPermittedForFileHierarchy(fs, fileStatus, userName, FsAction.READ, recurse)) {
+    if (FileUtils.isActionPermittedForFileHierarchy(fs, fileStatus, userName, FsAction.READ, recurse, fsAsUser)) {
       privs.add(SQLPrivTypeGrant.SELECT_NOGRANT);
     }
+    if (proxyUser != null) {
+      FileSystem.closeAllForUGI(proxyUser);
```

Comment:
Should not it be done for failure cases also ?

Reply:
Yes I agree, Updated